File: ios/Nexo/Features/Schedule/ScheduleViewModel.swift
//
//  ScheduleViewModel.swift
//  Nexo
//

import Foundation

@MainActor
@Observable
final class ScheduleViewModel {
    private let backend: BackendService
    let gymId: UUID

    var classes: [GymClass] = []
    var bookedClassIds: Set<UUID> = []
    var waitlistedClassIds: Set<UUID> = []
    /// The signed-in user's position in a given class's waitlist — populated
    /// per-row by `loadRowDetails(for:canManage:)` as each `ClassRow` appears,
    /// rather than upfront for the whole day (mirrors `ClassDetailViewModel`'s
    /// same fetch, just scoped to one row instead of one screen).
    var waitlistPositions: [UUID: Int] = [:]
    /// Checked-in attendee count per class, staff-only — same per-row
    /// loading rationale as `waitlistPositions`.
    var checkedInCounts: [UUID: Int] = [:]
    var isLoading = false
    var errorMessage: String?
    var selectedDate: Date = .now
    /// Non-nil signals a booking-action failure the view should present as an alert.
    var bookingMessage: String?
    /// Non-nil signals a successful booking the view should present as a
    /// transient, non-blocking toast — cleared by the view immediately after
    /// it's shown, mirroring how `bookingMessage` is cleared by its alert's
    /// dismiss binding, so back-to-back bookings of the same class each
    /// still trigger a fresh toast.
    var bookingSuccessMessage: String?
    /// Non-nil signals a successful waitlist join the view should present as
    /// a transient "Waitlisted!" toast — same pattern as `bookingSuccessMessage`,
    /// kept as a separate property so the view can tell the two outcomes apart
    /// and show the right title/icon.
    var waitlistSuccessMessage: String?
    /// This member's active wallet items for this gym — loaded once
    /// alongside booking/waitlist status, used by `bookingBlockedReason(for:)`
    /// to proactively dim the Book button before a doomed attempt is even
    /// made. Owners/Coaches/Platform Admins bypass this check entirely (the
    /// view skips calling `bookingBlockedReason` for them via
    /// `appState.canManageClasses`), so it doesn't matter that this is
    /// fetched for them too.
    var activePlans: [ActivePlanItem] = []

    private var stopListener: (() -> Void)?

    init(gymId: UUID, backend: BackendService? = nil) {
        self.gymId = gymId
        self.backend = backend ?? FirebaseBackend.shared
    }

    // MARK: - Data

    func loadInitialData() async {
        isLoading = true
        await loadBookingStatus()
        startObserving()
    }

    func dateChanged() {
        classes = []
        isLoading = true
        stopObserving()
        Task {
            await loadBookingStatus()
            startObserving()
        }
    }

    func loadBookingStatus() async {
        do {
            async let bookings = backend.fetchUserBookings(gymId: gymId)
            async let waitlist = backend.fetchUserWaitlist(gymId: gymId)
            bookedClassIds = (try? await bookings) ?? []
            waitlistedClassIds = (try? await waitlist) ?? []
        } catch {
            errorMessage = "Error loading bookings: \(error.localizedDescription)"
        }
        if let uid = backend.currentUID() {
            activePlans = (try? await backend.fetchActivePlans(gymId: gymId, userId: uid)) ?? []
        }
    }

    /// nil when this member has an active plan/credit balance covering
    /// `gymClass` (or when the caller bypasses this check — gated via
    /// `appState.canManageClasses`); otherwise a short reason to show in
    /// place of a dimmed Book button. Mirrors
    /// `ClassDetailViewModel.bookingBlockedReason`.
    func bookingBlockedReason(for gymClass: GymClass) -> String? {
        let matching = activePlans.filter { $0.matches(gymClass: gymClass) }
        if matching.contains(where: { $0.type == .unlimited || $0.availableCredits() > 0 }) {
            return nil
        }
        return matching.isEmpty ? "No active plan" : "No credits remaining"
    }

    func startObserving() {
        stopListener = backend.observeClasses(gymId: gymId, for: selectedDate) { [weak self] updatedClasses in
            self?.classes = updatedClasses.sorted { $0.startTime < $1.startTime }
            self?.isLoading = false
        }
    }

    func stopObserving() {
        stopListener?()
        stopListener = nil
    }

    /// Fetches this row's waitlist position (if the user is waitlisted for
    /// it) and checked-in count (if [canManage]) — called from `ClassRow`'s
    /// `.task(id:)` as each row appears, so the Schedule list can show the
    /// same "Attendees: (x/y) · Waitlist: (...) · Checked In: (...)" text as
    /// `ClassDetailView` without fetching this for every class up front.
    func loadRowDetails(for gymClass: GymClass, canManage: Bool) async {
        if waitlistedClassIds.contains(gymClass.id) {
            if let result = (try? await backend.fetchWaitlistPosition(gymId: gymId, classId: gymClass.id)) ?? nil {
                waitlistPositions[gymClass.id] = result.position
            }
        }
        if canManage {
            if let attendees = try? await backend.fetchAttendees(gymId: gymId, classId: gymClass.id) {
                checkedInCounts[gymClass.id] = attendees.filter { $0.isCheckedIn }.count
            }
        }
    }

    // MARK: - Actions

    /// Mirrors the backend's own `FBError.classInPast`/`MockBackendError.classInPast`
    /// gating — checked here too so the UI fails fast with a clear message instead
    /// of relying solely on a round-trip to the backend to reject the action.
    private func guardNotPast(_ gymClass: GymClass) -> Bool {
        guard gymClass.startTime < Date() else { return true }
        bookingMessage = "Cannot book or join waitlist for a class that has already started."
        return false
    }

    /// The live `observeClasses` listener eventually reconciles `classes` with the
    /// backend anyway, but that round-trip isn't instant — mutating the matching
    /// entry locally right after a successful action keeps the visible capacity/
    /// waitlist counts in sync immediately, without waiting on the listener.
    private func updateLocalClass(_ classId: UUID, _ transform: (inout GymClass) -> Void) {
        guard let index = classes.firstIndex(where: { $0.id == classId }) else { return }
        transform(&classes[index])
    }

    func bookClass(_ gymClass: GymClass) async {
        guard guardNotPast(gymClass) else { return }

        // Optimistic update
        let previousPlans = activePlans
        activePlans = consumeCreditLocally(activePlans, for: gymClass)
        bookedClassIds.insert(gymClass.id)
        updateLocalClass(gymClass.id) { $0.currentAttendees += 1 }
        bookingSuccessMessage = "\(gymClass.title) · \(gymClass.formattedTime)"

        do {
            try await backend.book(gymId: gymId, classId: gymClass.id)
        } catch {
            // Revert on failure
            activePlans = previousPlans
            bookedClassIds.remove(gymClass.id)
            updateLocalClass(gymClass.id) { $0.currentAttendees = max(0, $0.currentAttendees - 1) }
            bookingSuccessMessage = nil
            bookingMessage = "Failed to book: \(error.localizedDescription)"
        }
    }

    func cancelBooking(_ gymClass: GymClass) async {
        guard guardNotPast(gymClass) else { return }

        // Optimistic update
        let previousPlans = activePlans
        activePlans = refundCreditLocally(activePlans, for: gymClass)
        bookedClassIds.remove(gymClass.id)
        updateLocalClass(gymClass.id) { $0.currentAttendees = max(0, $0.currentAttendees - 1) }

        do {
            try await backend.cancelBooking(gymId: gymId, classId: gymClass.id)
        } catch {
            // Revert on failure
            activePlans = previousPlans
            bookedClassIds.insert(gymClass.id)
            updateLocalClass(gymClass.id) { $0.currentAttendees += 1 }
            bookingMessage = "Failed to cancel: \(error.localizedDescription)"
        }
    }

    func joinWaitlist(_ gymClass: GymClass) async {
        guard guardNotPast(gymClass) else { return }

        // Optimistic update
        waitlistedClassIds.insert(gymClass.id)
        updateLocalClass(gymClass.id) { $0.waitlistCount += 1 }
        waitlistSuccessMessage = "\(gymClass.title) · \(gymClass.formattedTime)"

        do {
            try await backend.joinWaitlist(gymId: gymId, classId: gymClass.id)
        } catch {
            // Revert on failure
            waitlistedClassIds.remove(gymClass.id)
            updateLocalClass(gymClass.id) { $0.waitlistCount = max(0, $0.waitlistCount - 1) }
            waitlistSuccessMessage = nil
            bookingMessage = "Failed to join waitlist: \(error.localizedDescription)"
        }
    }

    func leaveWaitlist(_ gymClass: GymClass) async {
        guard guardNotPast(gymClass) else { return }

        // Optimistic update
        waitlistedClassIds.remove(gymClass.id)
        updateLocalClass(gymClass.id) { $0.waitlistCount = max(0, $0.waitlistCount - 1) }

        do {
            try await backend.leaveWaitlist(gymId: gymId, classId: gymClass.id)
        } catch {
            // Revert on failure
            waitlistedClassIds.insert(gymClass.id)
            updateLocalClass(gymClass.id) { $0.waitlistCount += 1 }
            bookingMessage = "Failed to leave waitlist: \(error.localizedDescription)"
        }
    }

    private func consumeCreditLocally(_ plans: [ActivePlanItem], for gymClass: GymClass) -> [ActivePlanItem] {
        if plans.contains(where: { $0.matches(gymClass: gymClass) && $0.type == .unlimited }) { return plans }
        var list = plans
        if let targetIndex = list.firstIndex(where: { $0.matches(gymClass: gymClass) && $0.type == .credits && $0.availableCredits() > 0 }) {
            var item = list[targetIndex]
            if item.resetPeriod == .monthly {
                item.cycleCreditsUsed += 1
            } else {
                item.remainingCredits = max(0, item.remainingCredits - 1)
            }
            list[targetIndex] = item
        }
        return list
    }

    private func refundCreditLocally(_ plans: [ActivePlanItem], for gymClass: GymClass) -> [ActivePlanItem] {
        if plans.contains(where: { $0.matches(gymClass: gymClass) && $0.type == .unlimited }) { return plans }
        var list = plans
        if let targetIndex = list.firstIndex(where: { $0.matches(gymClass: gymClass) && $0.type == .credits }) {
            var item = list[targetIndex]
            if item.resetPeriod == .monthly {
                item.cycleCreditsUsed = max(0, item.cycleCreditsUsed - 1)
            } else {
                item.remainingCredits += 1
            }
            list[targetIndex] = item
        }
        return list
    }
}

File: android/app/src/main/java/com/nexo/app/ui/schedule/ScheduleViewModel.kt
package com.nexo.app.ui.schedule

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nexo.app.data.repository.BackendRepository
import com.nexo.app.domain.model.ActivePlanItem
import com.nexo.app.domain.model.GymClass
import com.nexo.app.domain.model.PlanComponentType
import com.nexo.app.domain.model.PlanResetPeriod
import com.nexo.app.domain.model.isOnLocalDate
import com.nexo.app.domain.model.weekDatesFor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate

/** Coordination logic for the Schedule/booking screen — kept out of the Composable per CLAUDE.md's "business logic stays out of views" rule. */
class ScheduleViewModel(
    private val repository: BackendRepository,
    private val gymId: String
) : ViewModel() {

    data class UiState(
        val isLoading: Boolean = true,
        val selectedDate: LocalDate = LocalDate.now(),
        val allClasses: List<GymClass> = emptyList(),
        val bookedClassIds: Set<String> = emptySet(),
        val waitlistedClassIds: Set<String> = emptySet(),
        /** The signed-in user's position in a given class's waitlist — populated per-row by [loadRowDetails] as each `ClassRow` appears, mirroring iOS's `ScheduleViewModel.waitlistPositions`. */
        val waitlistPositions: Map<String, Int> = emptyMap(),
        /** Checked-in attendee count per class, staff-only — same per-row loading rationale as [waitlistPositions]. */
        val checkedInCounts: Map<String, Int> = emptyMap(),
        /** Non-null once a Book/Waitlist action completes (optimistically, on tap) — the view shows the success popup then clears this via [clearSuccessMessage]. */
        val successMessage: SuccessMessage? = null,
        /** This member's active wallet items for this gym — loaded once alongside booked/waitlisted IDs, used by [bookingBlockedReason] to proactively dim the Book button before a doomed attempt is even made. Owners/Coaches/Platform Admins bypass this check entirely (the screen skips calling [bookingBlockedReason] for them), so it doesn't matter that this is fetched for them too. */
        val activePlans: List<ActivePlanItem> = emptyList(),
        val errorMessage: String? = null
    ) {
        val weekDates: List<LocalDate> get() = weekDatesFor(selectedDate)

        /** Classes from the live [BackendRepository.observeClasses] listener (all dates), filtered client-side to [selectedDate]. */
        val classesForSelectedDate: List<GymClass>
            get() = allClasses.filter { isOnLocalDate(it.startTimeMillis, selectedDate) }.sortedBy { it.startTimeMillis }

        /** null when this member can book [gymClass] (or bypasses the check — gated via `canManage`); otherwise a short reason to show in place of a dimmed Book button. Mirrors iOS's `ScheduleViewModel.bookingBlockedReason(for:)`. */
        fun bookingBlockedReason(gymClass: GymClass): String? {
            val matching = activePlans.filter { it.matches(gymClass) }
            if (matching.any { it.type == PlanComponentType.UNLIMITED || it.availableCredits() > 0 }) return null
            return if (matching.isEmpty()) "No active plan" else "No credits remaining"
        }
    }

    data class SuccessMessage(val title: String, val message: String, val isWaitlist: Boolean)

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val booked = repository.fetchMyBookedClassIds(gymId)
                val waitlisted = repository.fetchMyWaitlistedClassIds(gymId)
                _uiState.value = _uiState.value.copy(bookedClassIds = booked, waitlistedClassIds = waitlisted)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Error loading schedule: ${e.message}")
            }
            repository.currentUID()?.let { uid ->
                val plans = try { repository.fetchActivePlans(gymId, uid) } catch (e: Exception) { emptyList() }
                _uiState.value = _uiState.value.copy(activePlans = plans)
                viewModelScope.launch {
                    repository.observeActivePlans(gymId, uid).collect { livePlans ->
                        _uiState.value = _uiState.value.copy(activePlans = livePlans)
                    }
                }
            }
            observeClasses()
        }
    }

    /**
     * Subscribes to [BackendRepository.observeClasses] for the lifetime of
     * this ViewModel — mirrors iOS's `startObserving()`. Android doesn't
     * need an explicit `stopObserving()` teardown: `viewModelScope` is
     * cancelled automatically in `onCleared()`, which cancels this
     * collection (and, on Firebase, removes the underlying snapshot
     * listener via `awaitClose`) along with it. `isLoading` clears on the
     * first emission, same as iOS clearing it inside the listener callback.
     */
    private fun observeClasses() {
        viewModelScope.launch {
            repository.observeClasses(gymId).collect { classes ->
                _uiState.value = _uiState.value.copy(isLoading = false, allClasses = classes)
            }
        }
    }

    /** Silently re-reads booked/waitlisted IDs without showing a full-screen loading indicator (e.g. on resume after returning from ClassDetailScreen). Classes themselves stay live via [observeClasses] and don't need a manual refresh. */
    fun refresh() {
        viewModelScope.launch {
            try {
                val booked = repository.fetchMyBookedClassIds(gymId)
                val waitlisted = repository.fetchMyWaitlistedClassIds(gymId)
                _uiState.value = _uiState.value.copy(bookedClassIds = booked, waitlistedClassIds = waitlisted)
            } catch (e: Exception) {
                // Keep existing state on silent refresh error
            }
            repository.currentUID()?.let { uid ->
                try {
                    val plans = repository.fetchActivePlans(gymId, uid)
                    _uiState.value = _uiState.value.copy(activePlans = plans)
                } catch (e: Exception) {
                    // Keep existing state on silent refresh error
                }
            }
        }
    }

    fun selectDate(date: LocalDate) {
        _uiState.value = _uiState.value.copy(selectedDate = date)
    }

    /** Shifts [UiState.selectedDate] (and with it, the visible week strip) by [weeks] whole weeks — negative shifts backward. */
    fun shiftWeek(weeks: Long) {
        _uiState.value = _uiState.value.copy(selectedDate = _uiState.value.selectedDate.plusWeeks(weeks))
    }

    /**
     * Optimistic booking/waitlist actions — update [_uiState] immediately
     * (0ms perceived latency) instead of waiting on a network round-trip,
     * reverting on failure. Mirrors iOS's `ScheduleViewModel.bookClass`/
     * `cancelBooking`/`joinWaitlist`/`leaveWaitlist`.
     */
    fun cancel(classId: String) {
        val gymClass = _uiState.value.allClasses.firstOrNull { it.id == classId }
        val previousPlans = _uiState.value.activePlans
        val updatedPlans = if (gymClass != null) refundCreditLocally(previousPlans, gymClass) else previousPlans

        _uiState.value = _uiState.value.copy(
            bookedClassIds = _uiState.value.bookedClassIds - classId,
            activePlans = updatedPlans
        )
        updateLocalClass(classId) { it.copy(currentAttendees = (it.currentAttendees - 1).coerceAtLeast(0)) }

        viewModelScope.launch {
            try {
                repository.cancelBooking(gymId, classId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    bookedClassIds = _uiState.value.bookedClassIds + classId,
                    activePlans = previousPlans,
                    errorMessage = "Failed to cancel: ${e.message}"
                )
                updateLocalClass(classId) { it.copy(currentAttendees = it.currentAttendees + 1) }
            }
        }
    }

    fun leaveWaitlist(classId: String) {
        _uiState.value = _uiState.value.copy(waitlistedClassIds = _uiState.value.waitlistedClassIds - classId)
        updateLocalClass(classId) { it.copy(waitlistCount = (it.waitlistCount - 1).coerceAtLeast(0)) }

        viewModelScope.launch {
            try {
                repository.leaveWaitlist(gymId, classId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    waitlistedClassIds = _uiState.value.waitlistedClassIds + classId,
                    errorMessage = "Failed to leave waitlist: ${e.message}"
                )
                updateLocalClass(classId) { it.copy(waitlistCount = it.waitlistCount + 1) }
            }
        }
    }

    fun book(classId: String) {
        val gymClass = _uiState.value.allClasses.firstOrNull { it.id == classId }
        val previousPlans = _uiState.value.activePlans
        val updatedPlans = if (gymClass != null) consumeCreditLocally(previousPlans, gymClass) else previousPlans

        _uiState.value = _uiState.value.copy(
            bookedClassIds = _uiState.value.bookedClassIds + classId,
            activePlans = updatedPlans,
            successMessage = SuccessMessage(
                title = "Booked!",
                message = "${gymClass?.title.orEmpty()} · ${gymClass?.formattedTime.orEmpty()}",
                isWaitlist = false
            )
        )
        updateLocalClass(classId) { it.copy(currentAttendees = it.currentAttendees + 1) }

        viewModelScope.launch {
            try {
                repository.bookClass(gymId, classId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    bookedClassIds = _uiState.value.bookedClassIds - classId,
                    activePlans = previousPlans,
                    successMessage = null,
                    errorMessage = "Failed to book: ${e.message}"
                )
                updateLocalClass(classId) { it.copy(currentAttendees = (it.currentAttendees - 1).coerceAtLeast(0)) }
            }
        }
    }

    fun joinWaitlist(classId: String) {
        val gymClass = _uiState.value.allClasses.firstOrNull { it.id == classId }
        _uiState.value = _uiState.value.copy(
            waitlistedClassIds = _uiState.value.waitlistedClassIds + classId,
            successMessage = SuccessMessage(
                title = "Waitlisted!",
                message = "${gymClass?.title.orEmpty()} · ${gymClass?.formattedTime.orEmpty()}",
                isWaitlist = true
            )
        )
        updateLocalClass(classId) { it.copy(waitlistCount = it.waitlistCount + 1) }

        viewModelScope.launch {
            try {
                repository.joinWaitlist(gymId, classId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    waitlistedClassIds = _uiState.value.waitlistedClassIds - classId,
                    successMessage = null,
                    errorMessage = "Failed to join waitlist: ${e.message}"
                )
                updateLocalClass(classId) { it.copy(waitlistCount = (it.waitlistCount - 1).coerceAtLeast(0)) }
            }
        }
    }

    fun clearSuccessMessage() {
        _uiState.value = _uiState.value.copy(successMessage = null)
    }

    /**
     * Fetches this row's waitlist position (if the user is waitlisted for
     * it) and checked-in count (if [canManage]) — called from `ClassRow`
     * as each row appears, so the Schedule list can show the same
     * "Attendees: (x/y) · Waitlist: (...) · Checked In: (...)" text as
     * `ClassDetailScreen` without fetching this for every class up front.
     */
    fun loadRowDetails(classId: String, canManage: Boolean) {
        viewModelScope.launch {
            if (classId in _uiState.value.waitlistedClassIds) {
                val position = repository.fetchWaitlistPosition(gymId, classId)
                if (position != null) {
                    _uiState.value = _uiState.value.copy(waitlistPositions = _uiState.value.waitlistPositions + (classId to position))
                }
            }
            if (canManage) {
                val checkedInCount = repository.fetchAttendees(gymId, classId).count { it.isCheckedIn }
                _uiState.value = _uiState.value.copy(checkedInCounts = _uiState.value.checkedInCounts + (classId to checkedInCount))
            }
        }
    }

    private fun updateLocalClass(classId: String, transform: (GymClass) -> GymClass) {
        _uiState.value = _uiState.value.copy(
            allClasses = _uiState.value.allClasses.map { if (it.id == classId) transform(it) else it }
        )
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    private fun consumeCreditLocally(plans: List<ActivePlanItem>, gymClass: GymClass): List<ActivePlanItem> {
        if (plans.any { it.matches(gymClass) && it.type == PlanComponentType.UNLIMITED }) return plans
        val list = plans.toMutableList()
        val targetIndex = list.indexOfFirst {
            it.matches(gymClass) && it.type == PlanComponentType.CREDITS && it.availableCredits() > 0
        }
        if (targetIndex != -1) {
            val item = list[targetIndex]
            val updated = if (item.resetPeriod == PlanResetPeriod.MONTHLY) {
                item.copy(cycleCreditsUsed = item.cycleCreditsUsed + 1)
            } else {
                item.copy(remainingCredits = (item.remainingCredits - 1).coerceAtLeast(0))
            }
            list[targetIndex] = updated
        }
        return list
    }

    private fun refundCreditLocally(plans: List<ActivePlanItem>, gymClass: GymClass): List<ActivePlanItem> {
        if (plans.any { it.matches(gymClass) && it.type == PlanComponentType.UNLIMITED }) return plans
        val list = plans.toMutableList()
        val targetIndex = list.indexOfFirst {
            it.matches(gymClass) && it.type == PlanComponentType.CREDITS
        }
        if (targetIndex != -1) {
            val item = list[targetIndex]
            val updated = if (item.resetPeriod == PlanResetPeriod.MONTHLY) {
                item.copy(cycleCreditsUsed = (item.cycleCreditsUsed - 1).coerceAtLeast(0))
            } else {
                item.copy(remainingCredits = item.remainingCredits + 1)
            }
            list[targetIndex] = updated
        }
        return list
    }
}
